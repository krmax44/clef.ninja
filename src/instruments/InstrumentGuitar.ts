import Instrument from './Instrument';
import range from '@/utils/range';
import Note from '@/utils/Note';

const guitar: Instrument = {
	constants: {
		treble: {
			min: 55, // G3
			max: 85 // D6
		},

		bass: {
			min: 40, // E2
			max: 65 // F4
		}
	},
	clefs: ['treble', 'bass'],
	name: 'guitar'
};

type GuitarNote = { note: number; string: number; fret: number };

// see https://en.wikipedia.org/wiki/Guitar_tunings#Standard
// E2, A2, D3, G3, B3, E4
const BASENOTES = [40, 45, 50, 55, 59, 64];
const BASENOTESI = [...BASENOTES].reverse();

// fret widths - sums up to 100%
const FRETW = [1, 8, 8, 7, 7, 6, 6, 6, 6, 5, 5, 5, 4, 4, 4, 4, 3, 3, 3, 3, 2];

const FRETS = FRETW.map((width, fret) => {
	const strings = new Array(6).fill(undefined).map((a, string) => {
		const midi = BASENOTESI[string] + fret;
		return new Note(midi);
	});

	return { width, strings, fret };
});

const MIDIMAP = BASENOTES.map((basenote, string) => {
	const nextStop = BASENOTES[string + 1] || guitar.constants.treble.max;
	return [...range(basenote, nextStop)].map((note, fret) => ({
		string: 5 - string, // highest string is top here
		fret,
		note
	}));
}).flat();

function midiToGuitarNote(note: number): GuitarNote | undefined {
	return MIDIMAP.find(n => n.note === note);
}

export default guitar;
export {
	BASENOTES,
	BASENOTESI,
	FRETW,
	FRETS,
	MIDIMAP,
	GuitarNote,
	midiToGuitarNote
};
