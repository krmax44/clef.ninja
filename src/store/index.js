import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

const stages = ['gamemodeSelector', 'keyboardSetup', 'game'];

export default new Vuex.Store({
	state: {
		stage: 'gamemodeSelector',
		gamemode: '',
		keyboard: {
			type: 'virtual',
			midi: undefined
		}
	},
	mutations: {
		gamemode(state, gamemode) {
			state.gamemode = gamemode;
		},
		keyboard(state, keyboard) {
			state.keyboard = keyboard;
		},
		nextStage(state) {
			const currentStage = stages.indexOf(state.stage);
			state.stage = stages[currentStage + 1];
		},
		stage(state, stage) {
			state.stage = stages[stage];
		}
	}
});
