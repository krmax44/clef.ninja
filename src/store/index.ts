import Vue from 'vue';
import Vuex from 'vuex';
import InstrumentPiano from '@/instruments/InstrumentPiano';
import InstrumentGuitar from '@/instruments/InstrumentGuitar';

Vue.use(Vuex);

const instruments = {
	piano: InstrumentPiano,
	guitar: InstrumentGuitar
};

export default new Vuex.Store({
	state: {
		stage: 'homeView',
		gamemode: '',
		keyboard: {
			type: 'virtual',
			midi: undefined
		},
		instrument: InstrumentPiano
	},
	mutations: {
		gamemode(state, gamemode) {
			state.gamemode = gamemode;
		},
		keyboard(state, keyboard) {
			state.keyboard = keyboard;
		},
		stage(state, stage) {
			state.stage = stage;
		},
		instrument(state, instrument: string) {
			if (Object.keys(instruments).includes(instrument)) {
				state.instrument = instruments[instrument as 'guitar' | 'piano'];
			}
		}
	},
	getters: {
		midi(state) {
			return state.keyboard.type === 'midi' && state.keyboard.midi;
		}
	}
});
