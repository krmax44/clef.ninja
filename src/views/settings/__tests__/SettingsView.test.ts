import { shallowMount } from '@vue/test-utils';
import SettingsView from '../index.vue';
import InstrumentGuitar from '@/instruments/InstrumentGuitar';

import store from '@/store';

describe('SettingsView', () => {
	const wrapper = shallowMount(SettingsView, { store });

	it('sets the input type', () => {
		wrapper.find('.card-container').trigger('click');
		expect((wrapper.vm as any).$store.state.keyboard.type).toBe('virtual');
		expect((wrapper.vm as any).$store.state.stage.name).toBe('homeView');
	});

	it('sets the instrument', () => {
		wrapper
			.find('.setting:last-of-type .card-container:last-of-type')
			.trigger('click');
		expect((wrapper.vm as any).$store.state.instrument).toEqual('guitar');
	});

	it('renders correctly', () => {
		expect(wrapper.element).toMatchSnapshot();
	});
});
