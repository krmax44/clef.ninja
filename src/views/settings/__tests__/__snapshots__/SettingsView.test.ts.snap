// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsView renders correctly 1`] = `
<div
  class="px-4"
>
  <div
    class="setting"
  >
    <h2
      class="setting-title"
    >
      Input Setup
    </h2>
     
    <div
      class="flex justify-around py-8 flex-wrap"
    >
      <div
        class="card-container"
      >
        <div
          class="card"
        >
          <div
            class="card-inner"
          >
            <virtualicon-stub
              class="w-12 h-12"
              fillcolor="currentColor"
              size="24"
              title="Monitor Clean icon"
            />
             
            <h3>
              Virtual Keyboard
            </h3>
             
            <p>
              Use a virtual input device with your mouse.
            </p>
          </div>
        </div>
      </div>
       
      <div
        class="card-container"
      >
        <div
          class="card cursor-auto"
        >
          <div
            class="card-inner"
          >
            <midiicon-stub
              class="w-12 h-12"
              fillcolor="currentColor"
              size="24"
              title="Midi Port icon"
            />
             
            <h3>
              MIDI Keyboard
            </h3>
             
            <div>
              <p>
                
								Unfortunately, only Chromium-based browsers support this
								feature at the moment. We're sorry.
							
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="setting"
  >
    <h2
      class="setting-title"
    >
      Instrument
    </h2>
     
    <div
      class="flex justify-around py-8 flex-wrap"
    >
      <div
        class="card-container"
      >
        <div
          class="card flex"
        >
          <div
            class="card-inner"
          >
            <pianoicon-stub
              class="w-12 h-12"
              fillcolor="currentColor"
              size="24"
              title="Piano icon"
            />
             
            <h3>
              Piano
            </h3>
          </div>
        </div>
      </div>
       
      <div
        class="card-container"
      >
        <div
          class="card"
        >
          <div
            class="card-inner"
          >
            <guitaricon-stub
              class="w-12 h-12"
              fillcolor="currentColor"
              size="24"
              title="Guitar Acoustic icon"
            />
             
            <h3>
              Guitar
            </h3>
             
            <p>
              <span
                class="font-bold"
              >
                Beta:
              </span>
               This feature is a
							work-in-progress
						
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
